src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptCompletionServiceWithProviders.cs

```diff
@@ -22,25 +23,94 @@ internal sealed override CompletionRules GetRules(CompletionOptions options)
 
         internal abstract CompletionRules GetRulesImpl();
 
+        private static readonly ObjectPool<List<VSTypeScriptCompletionItemMatchResult>> s_listOfTSMatchResultPool = new(factory: () => new());
+        private static readonly ObjectPool<List<MatchResult>> s_listOfMatchResultPool = new(factory: () => new());
+
         internal sealed override void FilterItems(
            Document document,
            IReadOnlyList<MatchResult> matchResults,
            string filterText,
            IList<MatchResult> builder)
-            => FilterItemsImpl(document, matchResults, filterText, builder);
+        {
+            var tsMatchResults = s_listOfTSMatchResultPool.Allocate();
+            var tsFilteredMatchResults = s_listOfTSMatchResultPool.Allocate();
+
+            try
+            {
+                tsMatchResults.AddRange(matchResults.Where(static r => r.ShouldBeConsideredMatchingFilterText)
+                    .Select(static r => new VSTypeScriptCompletionItemMatchResult(r)));
+
+                FilterItemsImpl(document, tsMatchResults, filterText, tsFilteredMatchResults);
+                builder.AddRange(tsMatchResults.Select(static r => r.MatchResult));
+            }
+            finally
+            {
+                // Don't call ClearAndFree, which resets the capacity to a default value.
+                tsMatchResults.Clear();
+                tsFilteredMatchResults.Clear();
+                s_listOfTSMatchResultPool.Free(tsMatchResults);
+                s_listOfTSMatchResultPool.Free(tsFilteredMatchResults);
+            }
+        }
 
+        // Override this to bypass the expensive call to public FilterItems method.
         internal virtual void FilterItemsImpl(
```

Comment:
tbh, i'm finding the hierarchy here of FilterItems and FilterItemsImpl to be very difficult ot reason about.

Reply:
do we need the two virtual FilterItems on CompletionService (neither of which seem to call each other??).

Reply:
the one overload we are trying to avoid is public so it has to be preserved, and we can't make the internal one public to replace it yet (w/o exposing a bunch of things we don't want to).